### ISSUE TITLE
Child Nodes receive an outdated version of the schema. 

### ISSUE THREAD

**Issue by volkan-a** [antonioru/beautiful-react-diagrams#6]
I'm not sure if it is the correct way but I want to use custom node renderer like child component. I want to manipulate schema object when I click on custom node. But I'm just getting (probably) a copy of schema state for each node.

If I try the code below and add a few nodes, on each click I see different length of schema.nodes? How can I get "correct" schema?

```javascript
import Diagram from 'beautiful-react-diagrams';

const CustomNode = (props) => {
    const {
        inputs
    } = props;
    const {
        onClick
    } = props.data;

    return ( <
        div style = {
            {
                background: '#717EC3',
                borderRadius: '10px'
            }
        }
        onClick = {
            onClick
        } >
        <
        div style = {
            {
                padding: '10px',
                color: 'white'
            }
        } >
        Custom Node <
        /div> <
        div style = {
            {
                marginTop: '20px'
            }
        } > {
            inputs.map((port) => React.cloneElement(port, {
                style: {
                    width: '50px',
                    height: '25px',
                    background: '#1B263B'
                }
            }))
        } <
        /div> <
        /div>
    );
};

const initialSchema = {
    nodes: [{
            id: 'node-1',
            content: 'Node 1',
            coordinates: [150, 60],
            outputs: [{
                id: 'port-1',
                alignment: 'right'
            }],
        },
        {
            id: 'node-custom',
            coordinates: [250, 60],
            render: CustomNode,
            inputs: [{
                id: 'custom-port-1',
                alignment: 'left'
            }],
        },
    ]
};

const UncontrolledDiagram = () => {
    // create diagrams schema
    const [schema, setSchema] = React.useState(initialSchema);

    const showInfo = () => {
        console.log(schema.nodes.length);
    }

    const addNode = () => {
        const node = {
            id: Math.random().toString(),
            coordinates: [250, 60],
            render: CustomNode,
            data: {
                onClick: showInfo,
            },
            inputs: [{
                id: 'custom-port-1',
                alignment: 'left'
            }],
        }
        setSchema({
            ...schema,
            nodes: [...schema.nodes, node]
        });
    }

    return ( <
        div style = {
            {
                height: '22.5rem'
            }
        } >
        <
        button onClick = {
            () => addNode()
        } > Add < /button> <
        Diagram schema = {
            schema
        }
        onChange = {
            setSchema
        }
        /> <
        /div>
    );
};

<
UncontrolledDiagram / >
```

**Comment by antonioru** [comment]
Hello @volkan-a 

thank you for your question, I hope I can be of any help. 
Can you please provide a [CodeSandbox](https://codesandbox.io/) of your code? 
That would be much easier to debug

**Comment by volkan-a** [comment]
https://3tlfy.csb.app

It seem a bit ugly but it helps. First add  a few custom nodes then click on them. You see different length of schema.nodes on console.

**Comment by antonioru** [comment]
If I'm getting this right, the problem is that the child-nodes are receiving a copy of the schema from the moment they're created. 
For this reason you'll have different schema in different moment in time. 

I think this is a bug, feel free to open a PR if you can/want

**Comment by antonioru** [comment]
@volkan-a 

I've dug further into this and it seems the problem lay somewhere into the fact the state is a cloned-object but when the Diagram component receives it's already a copy, I'm not sure how to properly solve this... do you have something in mind?

I think a pair of new hooks built on top of `useReducer` would help, something like:


```
const [schema, schemaMethods] = useSchema(); 

const addNode = () => schemaMethods.addNode({
    coordinates: [10, 10],
    content: 'Hello world',
    ...others,
});

return (
         <div>
              <Button onClick={addNode}>Add node</Button>
             <Diagram schema={schema) onChange={schemaMethods.onChange} />
        </div>
);
```

**Comment by antonioru** [comment]
All right, the first experimental implementation of `useSchema` is available at [feature/useSchema](https://github.com/antonioru/beautiful-react-diagrams/tree/feature/useSchema) branch! 

The Diagram onChange method signature is changed so from this feature on you **need** to use `useSchema`!  
This would fix this bug, please feel free to give any feedback.

**Comment by antonioru** [comment]
Closed by #24
